File: 소프티어/Lv3/수퍼바이러스.py
import sys
input = sys.stdin.readline

K, P, N = map(int, input().split())
N *= 10

def check(x, y):
    if y == 1:
        return x
    elif y % 2 == 1:
        res = check(x, (y - 1) / 2)
        return res * res * x % 1000000007
    else:
        res = check(x, y / 2)
        return res * res % 1000000007

ans = check(P, N) * K
print(ans % 1000000007)

File: 소프티어/JavaScript/Lv3/수퍼바이러스.js
const readline = require('readline')
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
})

let input;

rl.on('line', function(line){
    input = line.split(' ').map(Number)
}).on('close', function() {
    let K = BigInt(input[0])
    const P = BigInt(input[1])
    const N = BigInt(input[2] * 10)

    // console.log(K, P, N)

    const check = (x, y) => {
        if (y === BigInt(1)){
            return BigInt(x)
        } else if (y % BigInt(2) === BigInt(1)){
            const res = check(x, (y - BigInt(1)) / BigInt(2))
            return res * res * x % BigInt(1000000007)
        } else{
            const res = check(x, y / BigInt(2))
            return res * res % BigInt(1000000007)
        }
    }

    ans = check(P, N) * K
    ans %= BigInt(1000000007)

    console.log(Number(ans))
})